js/bittrex.js

```diff
@@ -3,7 +3,7 @@
 //  ---------------------------------------------------------------------------
 
 const Exchange = require ('./base/Exchange');
-const { BadSymbol, ExchangeError, ExchangeNotAvailable, AuthenticationError, InvalidOrder, InsufficientFunds, OrderNotFound, DDoSProtection, PermissionDenied, AddressPending } = require ('./base/errors');
+const { BadSymbol, ExchangeError, ExchangeNotAvailable, AuthenticationError, InvalidOrder, InsufficientFunds, OrderNotFound, DDoSProtection, PermissionDenied, AddressPending, MaintananceError } = require ('./base/errors');
```

Comment:
We have `ExchangeNotAvailable` which may also be thrown in case of unavailability due to maintenance... I'm think if we should me this type of error a subtype of `ExchangeNotAvailable`... what do you think? Something like `ExchangeNotAvailable → OnMaintenance`?

Reply:
@kroitor 
I looked through the existing exceptions. I think, the problem with `ExchangeNotAvailable` is, that it is a `NetworkError`, which is not actually the case. `OnMaintenance` sounds good to me, I can adopt the PR, when we agreed upon a good structure.

Reply:
> I think, the problem with ExchangeNotAvailable is, that it is a NetworkError, which is not actually the case. 

That's a good point, and there isn't a definitive answer on that, it depends on how you define your "axioms".

If you treat a `NetworkError` like a low-level hardware/software connection error (IP-level issues, host unreachable, DNS issues, ddos protections of all sorts), then the `OnMaintenance` exception does not seem to be a kind of `NetworkError`, of course. However, if we're too strict with our definitions and especially picky on words – the `OnMaintenance` isn't an `ExchangeError` either, and in fact, isn't an error at all, since being on maintenance is a normal mode of operation, not a parameter error, nor a crash on backend, nor a bug in ccxt, so, not an error. 

So, if we try to be too rigid with our definition of the error hierarchy, then we will have hard time defining the hierarchy, because in many cases it's very hard to define precise boundaries, purpose and meaning, say, a Parameter Error can be both `InvalidOrder` and/or `BadRequest`. So, the actual goal behind the hierarchy of errors is to have as few specific exceptions as possible – to keep it maintainable, that way we won't have to break our heads on trying to strictly define how those exceptions relate to each other. Another design goal is to put as few catch-clauses on the user side as possible (this one is obvious). The exchanges themselves only make it worse – some of them will report maintenance as connectivity errors...

Therefore I suggest that we treat the error hierarchy loosely, for example, I propose the following flexible definition of a `NetworkError`:

_A `NetworkError` is not an error, it's a temporary unavailability situation, which may be caused by any condition or by any type of availability factor, including maintenance, ddos protections and temporary bans. The reason for having a big family of `NetworkErrors` is to group all exceptions that can reappear or disappear upon a later retry or a retry from a different location, all the rest being equal (with the same user input, without any changes, basically, same order price and amount, same symbol, etc...)._

_In contrast, the `ExchangeError` is an error indeed, and it differs from the `NetworkError` in a very specific way – if you get an `ExchangeError` with your input, then you should always get the same `ExchangeError` with that same input._

In other words, the more formal distinction between the two families of exchanges: one is recoverable (`NetworkError` = can retry later and it can magically go away, so a subsequent retry may succeed) and the other is not (`ExchangeError` = is a fatal error, so something went bad and it will go bad every time, unless you change the input).

If you apply that loose definition of the two groups, and forget for a second about the word `Error` in the `NetworkError` family, then `OnMaintenance` falls perfectly into that group, because it is a recoverable error – can go away by itself some time later. Suppose that under `NetworkError` we assume _`SomethingIsWrongTemporarily`_ so:

`NetworkError (SomethingIsWrongTemporarily) → ExchangeNotAvailable → OnMaintenance`

When the exchange is on maintenance and you get an error response – that is definitely not a problem with your input and not a problem with the exchange engine, and in that regard, it is very unlikely an `ExchangeError`. 

The more general way to think about the two families is:

- NetworkError = any condition when the clientside is encouraged to retry later
- ExchangeError = any condition when the clientside is required to interfere, stop and investigate (or change tactics on the fly)

Does the above make sense?

Reply:
Great arguments, thank you very much for taking time for this detailed explanation, maybe you can copy&paste some of the above in the documentation?
I totally agree with your arguments.
My thought of not having it as a NetworkError was based on my userland point of view:
I catch the NetworkErrors and simply retry some seconds later, which normally works then. (it is a bit more complicated, waiting longer if the same error occurs for example).
A maintenance usually talks longer (some hours for bitfinex/bittrex for example), so waiting longer per default for OnMaintanance makes sense, but if I rethink this, this is not always the case and I also can change the behaviour for OnMaintenance if it is a NetworkError.
My distinction was based on: fast recovery vs slow recovery, but since ExchangeError should never recover, this is ok for me and fits even better.

Reply:
> maybe you can copy&paste some of the above in the documentation?

Sure, will do asap! 

The wiki is editable btw, because it's in the same repo:

- https://github.com/ccxt/ccxt/tree/master/wiki
- https://github.com/ccxt/ccxt/blob/master/wiki/Manual.md

Your help is always welcome, if you notice that something is missing there or if some of it can be explained in better ways – plz don't hesitate, your participation has been invaluable!) Grammar nazi are also welcome )

Reply:
@brandsimon one specific point i totally agree on is that the naming may be a bit confusing, but this is for historical reasons mostly, from the early adolescent days of the project... We'd probably name them `TemporaryError` and `PermanentError`, or close to that, but now it's too late ))

Reply:
@kroitor 
Great to hear!
Thank you for finishing this PR.
Should we make a smooth transition of the Exceptions then?

1. rename them, but keep the old names around. (`NetworkError = TemporaryError` should do in all of the languages)
2. Deprecate them (put a warning in their Constructor)
3. Remove them

Reply:
@brandsimon yep, i'm all for it, will get this migration started!

Reply:
@kroitor 
Great! Let me know if you need help with anything.

Is there a way to stay informed about breaking changes?
Easiest way is probably to use github: One Issue, where people can subscribe and every new change (a separate issue) is listed there, so people will get notified.

Reply:
> One Issue, where people can subscribe

Yep, the only difficulty is maintaining to keep it up )) It's particularly hard to find the time on micro-management, but i hope to approach it again soon after the release of CCXT Pro on which we currently spend most of our time.